File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/TypeDamageRelationServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation;
import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation.TypeDamageRelationId;
import io.github.lishangbu.avalon.dataset.repository.TypeDamageRelationRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class TypeDamageRelationServiceImplTest {
    @Mock private TypeDamageRelationRepository typeDamageRelationRepository;
    @InjectMocks private TypeDamageRelationServiceImpl typeDamageRelationService;

    @Test
    void getPageByCondition_callsRepository() {
        TypeDamageRelation relation = new TypeDamageRelation();
        Pageable pageable = PageRequest.of(0, 5);
        Page<TypeDamageRelation> page = new PageImpl<>(List.of(relation));
        when(typeDamageRelationRepository.findAll(any(Example.class), eq(pageable)))
                .thenReturn(page);

        Page<TypeDamageRelation> result =
                typeDamageRelationService.getPageByCondition(relation, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(typeDamageRelationRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void listByCondition_allowsNullProbe() {
        List<TypeDamageRelation> expected = List.of();
        when(typeDamageRelationRepository.findAll(any(Example.class))).thenReturn(expected);

        List<TypeDamageRelation> result = typeDamageRelationService.listByCondition(null);

        assertSame(expected, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(typeDamageRelationRepository).findAll(captor.capture());
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void save_usesRepository() {
        TypeDamageRelation relation = new TypeDamageRelation();
        when(typeDamageRelationRepository.save(relation)).thenReturn(relation);

        TypeDamageRelation result = typeDamageRelationService.save(relation);

        assertSame(relation, result);
        verify(typeDamageRelationRepository).save(relation);
    }

    @Test
    void update_usesRepository() {
        TypeDamageRelation relation = new TypeDamageRelation();
        when(typeDamageRelationRepository.save(relation)).thenReturn(relation);

        TypeDamageRelation result = typeDamageRelationService.update(relation);

        assertSame(relation, result);
        verify(typeDamageRelationRepository).save(relation);
    }

    @Test
    void removeById_callsRepository() {
        TypeDamageRelationId id = new TypeDamageRelationId();
        typeDamageRelationService.removeById(id);
        verify(typeDamageRelationRepository).deleteById(id);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/TypeDamageRelationServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation
import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelationId
import io.github.lishangbu.avalon.dataset.repository.TypeDamageRelationRepository
import org.babyfish.jimmer.Page
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.data.domain.PageRequest

@ExtendWith(MockitoExtension::class)
class TypeDamageRelationServiceImplTest {
    @Mock
    private lateinit var typeDamageRelationRepository: TypeDamageRelationRepository

    @InjectMocks
    private lateinit var typeDamageRelationService: TypeDamageRelationServiceImpl

    @Test
    fun getPageByCondition_callsRepository() {
        val pageIndex = 0
        val pageSize = 5
        val pageable = PageRequest.of(pageIndex, pageSize)
        val page: Page<TypeDamageRelation> = Page(listOf(TypeDamageRelation()), 1, 1)
        Mockito.`when`(
            typeDamageRelationRepository.findPage(1L, 2L, 2.0f, pageable),
        ).thenReturn(page)

        val result = typeDamageRelationService.getPageByCondition(1L, 2L, 2.0f, pageable)

        assertSame(page, result)
        Mockito.verify(typeDamageRelationRepository).findPage(1L, 2L, 2.0f, pageable)
    }

    @Test
    fun listByCondition_allowsNullFilter() {
        val expected = listOf<TypeDamageRelation>()
        Mockito.`when`(typeDamageRelationRepository.findAll(null, null, null)).thenReturn(expected)

        val result = typeDamageRelationService.listByCondition(null, null, null)

        assertSame(expected, result)
        Mockito.verify(typeDamageRelationRepository).findAll(null, null, null)
    }

    @Test
    fun save_usesRepository() {
        val relation = TypeDamageRelation()
        Mockito.`when`(typeDamageRelationRepository.save(relation)).thenReturn(relation)

        val result = typeDamageRelationService.save(relation)

        assertSame(relation, result)
        Mockito.verify(typeDamageRelationRepository).save(relation)
    }

    @Test
    fun update_usesRepository() {
        val relation = TypeDamageRelation()
        Mockito.`when`(typeDamageRelationRepository.save(relation)).thenReturn(relation)

        val result = typeDamageRelationService.update(relation)

        assertSame(relation, result)
        Mockito.verify(typeDamageRelationRepository).save(relation)
    }

    @Test
    fun removeById_callsRepository() {
        val id = TypeDamageRelationId()
        typeDamageRelationService.removeById(id)
        Mockito.verify(typeDamageRelationRepository).deleteById(id)
    }
}
